engine/src/simulation.cpp

```diff
@@ -1142,14 +1142,64 @@ StateId SimulationMachine::Reset::impl(SimulationContext& ctx) {
     }
     return true;
   });
+  ctx.sync.reset();
+  ctx.vehicles.clear();
+
+  // Create and configure all controllers:
+  for (const auto& c : ctx.config.vehicles) {
+    std::shared_ptr<cloe::Vehicle> v;
+    if (c.is_from_simulator()) {
+      auto& s = ctx.simulators.at(c.from_sim.simulator);
+      if (c.from_sim.is_by_name()) {
+        v = s->get_vehicle(c.from_sim.index_str);
+        if (!v) {
+          throw cloe::ModelError("simulator {} has no vehicle by name {}",
+                                 c.from_sim.simulator, c.from_sim.index_str)
+              .explanation("Simulator {} has following vehicles:\n{}",
+                           c.from_sim.simulator,
+                           enumerate_simulator_vehicles(*s));
+        }
+      } else {
+        v = s->get_vehicle(c.from_sim.index_num);
+        if (!v) {
+          throw cloe::ModelError("simulator {} has no vehicle at index {}",
+                                 c.from_sim.simulator, c.from_sim.index_num)
+              .explanation("Simulator {} has following vehicles:\n{}",
+                           c.from_sim.simulator,
+                           enumerate_simulator_vehicles(*s));
+        }
+      }
+      for (const auto& ctrl : ctx.config.controllers) {
+        if (ctrl.vehicle == c.name) {
+          auto name = ctrl.name.value_or(ctrl.binding);
+          assert(ctx.controllers.count(name) != 0);
+          ctx.controllers[name]->set_vehicle(v);
+        }
+      }
+      for (const auto& veh : ctx.config.vehicles) {
+        (void)veh;
+        // shall we reset ctx.vehicle from carmaker binding?
+      }
+    } else {
+      logger()->info("Shall we reset veh from vehicles ?");
+      //   if (ctx.vehicles.count(c.from_veh)) {
+      //     v = ctx.vehicles.at(c.from_veh);
+      //   } else {
+      //     // This vehicle depends on another that hasn't been create yet.
+      //     return nullptr;
+      //   }
+      // }
```

Comment:
- [ ] What are the full semantics of resetting a simulation?

We need a document defining exactly what happens during a reset:

- [ ] How are simulators reset?
- [ ] How are controllers reset?
- [ ] How are components reset?
- [ ] How are hidden components reset?
- [ ] How are vehicles reset?
- [ ] Is there any other data that needs to be reset?

Reply:
**_How are simulators reset?
How are controllers reset?_**
I think it highly depends on if simulator and controller support "RESET" semantics, cloe-engine provide reset API seems to be enough.

We already finish the reset in simulator and controller part, and go through about if there are other "hidden" components are out of consideration, luckily it is working now.  
https://inside-docupedia.bosch.com/confluence/pages/resumedraft.action?draftId=3734864150&draftShareId=9f1a3e8c-3344-4726-9894-5005896aafda&


**_What are the full semantics of resetting a simulation?_**
This is rather difficult to answer from user side , I think feature owners from DE platform must have more clear overview.

Reply:
These trigger topics definitely need to be fixed.

Other questions: how to we reset all the triggers correctly?

Reply:
What to do with the simulation report?
- Best to output an array of each report for each simulation so we can automatically compare simulation runs to ensure correctness.

Reply:
**_how to we reset all the triggers correctly_**
My idea was moving trigger from START state to CONNECT state, but I cannot evaluate the risk to entire simulation. From my understanding we do not need to manipulate trigger any more after reset. (I do not have much deep dive into trigger mechinism :))

Reply:
The open points aren't necessarily your responsibility, I'm just noting them for completeness. From my perspective these are the things we need to somehow answer if we want to have a robust, well-defined simulation.

However, I'm also open to have this feature "experimental" where not all these points are answered initially.

Reply:
For example, our initial implementation will have a very restricted "feature-set":

- Simulator scenaria stays the same
- Ego vehicle stays the same
- Vehicle configuration stays the same
- Random seed not reset to initial state
- ...

What do we do:
- "reset()" method called on all components of each vehicle
- "reset()" method called on all simulation plugins
- ...